compiler/rustc_mir/src/borrow_check/type_check/mod.rs

```diff
@@ -180,7 +179,66 @@ pub(crate) fn type_check<'mir, 'tcx>(
             liveness::generate(&mut cx, body, elements, flow_inits, move_data, location_table);
 
             translate_outlives_facts(&mut cx);
-            cx.opaque_type_values
+            let mut opaque_type_values = cx.opaque_type_values;
+
+            for (_, revealed_ty) in &mut opaque_type_values {
+                // FIXME(oli-obk): Instead of looping, implement a visitor like
+                // FullTypeResolver. We can't use FullTypeResolver here, as that will
+                // resolve lifetimes lexically, which it can't because we didn't do old
+                // borrowck stuff. We want to use MIR borrowck information instead.
+
+                while revealed_ty.has_infer_types_or_consts() {
+                    let prev = *revealed_ty;
+                    trace!(prev=?prev.kind());
+                    let type_resolved = infcx.shallow_resolve(prev);
```

Comment:
How does this work on types that have only nested inference variables? `shallow_resolve` only resolves inference variables at the top level, so a type that doesn't have an inference variable at the top level would break in this loop.

Reply:
You're right. I misunderstood what `shallow_resolve` does. I'll implement a visitor as the FIXME above suggests

Reply:
Turns out there already was a pre-existing visitor that did the right thing™️